Feature/py open ssl

SvenRoederer opened the pull request:
This is my idea to issue #14  - this is WIP and only for discussion, but generated certificates match currently used 
- uses PyOpenSSL to do the key-generation and Cert-signing 
- no use of "easyrsa" shell calls
- uniqueness of cert serial is assured by database 

please comment on this

SvenRoederer committed:
[buildcert] add basic cert-sign example
- taken from https://gist.github.com/ril3y/1165038

SvenRoederer committed:
[buildcert] mods for standalone-execution

SvenRoederer committed:
[buildcert] generate the same cert-subject as old-script

SvenRoederer committed:
[buildcert] rename function to create_cert()

SvenRoederer committed:
sign the cert by the correct issuer

SvenRoederer committed:
add the same cert extensions
- Basic Constraints
- Netscape Comment
- Subject Key Identifier
- Authority Key Identifier
- (Extended) Key Usage

SvenRoederer committed:
sign cert with ca-key

SvenRoederer committed:
reenable flask

SvenRoederer committed:
cleanup & integrate to current code
- put key-generation into separate function
- integrate cert-generation into working-loop
- remove unused code to frite files

SvenRoederer committed:
cleanup & integrate to current code
- put key-generation into separate function
- integrate cert-generation into working-loop
- remove unused code to frite files

SvenRoederer committed:
upgrade to python3

SvenRoederer committed:
unique SN per certificate
- store individual SN of certificate for each request (column cert_sn)
- as "autoincrement" columns are not avail on all dbms, use MAX() on the
  "cert_sn" + 1
- "cert_sn" is set to be unique, so not doulble use of SN possible

SvenRoederer committed:
Merge branch 'feature/pyOpenSSL' of https://github.com/SvenRoederer/freifunk-berlin-ca.berlin.freifunk.net into feature/pyOpenSSL

Conflicts:
	buildcert.py

SvenRoederer committed:
[model] add class-method "getMaxCertSn()"
- looks up the highest assigned Certificate-serialnumber

SvenRoederer commented:
is this idea to bad, that nobody likes to check on this?

booo commented:
We should nice up the code a bit but the direction is perfect!

I suggest that we generate the id for the request in the database. Otherwise we may run into issues with multiple process creating the same id. If we generate the id when the request enters the database we may waste some id space when we reject the request but who cares?

booo commented on buildcert.py:
```diff
@@ -7,16 +7,82 @@
+    """
+
+    # get required CA-data
+    ca_cert_file = open('/tmp/ffca.crt', 'r')
```
This should be part of config.py

booo commented on buildcert.py:
```diff
@@ -7,16 +7,82 @@
+        # create a self-signed cert
+        cert = crypto.X509()
+        cert.set_version(0x02)  # X509-Version 3
+        cert.get_subject().C = "DE"
```
This should be part of config.py too.

booo commented on buildcert.py:
```diff
@@ -7,16 +7,82 @@
+        cert.get_subject().emailAddress = cert_email
+        cert.set_serial_number(cert_sn)
+        cert.gmtime_adj_notBefore(0)
+        cert.gmtime_adj_notAfter(10*365*24*60*60)
```
Same here is true for notAfter

booo commented on buildcert.py:
```diff
@@ -7,16 +7,82 @@
+        ]
+        cert.add_extensions(cert_ext)
+
+        cert.sign(ca_key, 'sha1')
```
Can we just switch to sha2? If we want to use the certificates in the future that's maybe a "good" move.

booo commented on buildcert.py:
```diff
@@ -7,16 +7,82 @@
+        new_cert_sn = Request.getMaxCertSn() + 1
+        request.cert_sn = new_cert_sn
+        new_cert = create_cert(request.id, request.email, request.cert_sn, new_key)
+        print (crypto.dump_certificate(crypto.FILETYPE_TEXT, new_cert))
```
We should save the cert/key somewhere. I guess openvpn expects the keys at least for revocation to be somewhere.

booo commented:
If you move this to a branch inside the repository we can collaborate on this.

Thanks for the proposal. Sorry for the late replay here.

SvenRoederer commented:
- I agree that some setting have to move to config.py
- Cert-serial: check commit [model] add class-method "getMaxCertSn()" --> S/N is generated by DB or will at least cause a constraint violation. current "database-id" is the is-token entered on web-site. but per requirement the Certs-ID has to be numerical and of course unique.
- shall we store the keys and certs in the DB or in file-archives as we do now?
- for revocation we have to think of a CRL-infrastructure ... and the CRLs only need the final cert or just the SN.

SvenRoederer committed:
buildcert.py: add some documentation for new functions
- as complained by quantifiedcode.com test

booo commented:
There is a revocation script. Can you check what it does and  what openvpn expects? Maybe this will answer the question of where to store the keys/certs too.

SvenRoederer committed:
models.py: add some documentation for new functions
- as complained by quantifiedcode.com test

SvenRoederer committed:
config, buildcert: implement configurable parameters

just move all relevant paramters into config.py to have them not hardcoded

SvenRoederer committed:
correcting parameter-names

SvenRoederer committed:
config.py: remove leftovers from initial try to use python-dicts

SvenRoederer commented:
for revocation OpenVPN expects the regular CRLs as file on disk.
Genrating the CRLs can be done offline via openssl command or even integrated into the web-app (for downloading by cron on each VPN-server)

SvenRoederer commented:
uploaded an improved version (including use of config.py) to branch https://github.com/freifunk-berlin/ca.berlin.freifunk.net/tree/pyOpenSSL 

closing this RFC-PR for further work on [ 3505097 ]

SvenRoederer committed:
Merge branch 'wip/pyOpenSSL-config' into wip/pyOpenSSL-config_after-manage-rework

Conflicts:
	buildcert.py
	config.py

SvenRoederer committed:
Merge branch 'master' of https://github.com/freifunk-berlin/ca.berlin.freifunk.net into wip/pyOpenSSL-config_after-manage-rework

Conflicts:
	manage.py

SvenRoederer committed:
migration: import old Eay-RSA certificate-serials into database

this is done during migration via import_easyrsa_certsn(), by matching
each database-record with generated certificate to the easy-RSA index-file.
Then the Certificate-S/N will be copied

SvenRoederer committed:
manage: cleanup some comments

SvenRoederer committed:
templates-vpn03: added default files

these files will be put into the archive which we will send

SvenRoederer committed:
templates: move

SvenRoederer committed:
manage.py: integrate creation of cert-files and tar-archive

create functions cert_store() and cert_createTar()
cert_store(): stores the generate key and certificate on filesystem
              - location can be set via config.DIRECTORY
cert_createTar(): generates a tar-archive with the required files
                  - reads tempalte-fiels from ca/template/vpn03-files
                  - reads certificate and key from config.DIRECTORY_CLIENTS

the tar-archive is recreated everytime before the mail will be send

SvenRoederer committed:
manage.py: fix indents for create_key()

SvenRoederer committed:
Revert "migration: import old Eay-RSA certificate-serials into database"

This reverts commit 03bec515a913e3dd669eb5d88d830dfd095007f8.

SvenRoederer committed:
import.py: add code to import existing Easy-RSA certificate-serials

moved from migration-script

SvenRoederer committed:
config.py: changed location of ca-files to use

SvenRoederer committed:
import.py: reorder definiton of code and actually use the function

SvenRoederer committed:
manage.py: fix os.path.join() -->join() is already imported

SvenRoederer committed:
manage.py: fix storing certificate / key in files

- data needs to be converted to PEM-encoding
- binary filemode needed

SvenRoederer committed:
vpn03-templates: remove individual names from file-explanation

- this saves additional parsing
- the filenames speaking for themself

SvenRoederer committed:
vpn03-templates: remove deprecated .p12 for android

- this file is not supplied since some time

SvenRoederer committed:
manage.py: fix os.listdir()

SvenRoederer committed:
vpn03-templates: hardcoce "freifunk-ca.crt" to config-files

SvenRoederer committed:
manage.py: change filenames inside certificate tar-archive

prefix the certificate-files with VPN03 (hardcoded)

SvenRoederer commented:
@booo @KenTacki - please comment on this
- needs some cleanup + rebasing
- seems to work for me
- generates "dropin" certificates / as much compatibility as possible to current setup

SvenRoederer committed:
config.py: use relative path

SvenRoederer commented:
looks like it was auto-reopened
